src/Symfony/Component/String/GraphemeString.php

```diff
@@ -0,0 +1,341 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+class GraphemeString extends AbstractString
+{
+    public function __construct(string $string = '', string $locale = null)
+    {
+        $this->string = normalizer_normalize($string);
+
+        if (false === $this->string) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        if (null !== $locale) {
+            $this->locale = $this->withLocale($locale)->locale;
+        }
+    }
+
+    public function afterFirst(string $needle, bool $includeNeedle = false): parent
+    {
+        $str = clone $this;
+
+        if ('' === $needle) {
+            $str->string = '';
+        } elseif (false === $needle = normalizer_normalize($needle)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        } elseif ($this->ignoreCase) {
+            $i = grapheme_stripos($this->string, $needle);
+            $str->string = false === $i ? '' : grapheme_substr($str->string, $i + ($includeNeedle ? 0 : grapheme_strlen($needle)));
+        } else {
+            $i = strpos($this->string, $needle);
+            $str->string = false === $i ? '' : substr($str->string, $i + ($includeNeedle ? 0 : \strlen($needle)));
+        }
+
+        return $str;
+    }
+
+    public function afterLast(string $needle, bool $includeNeedle = false): parent
+    {
+        $str = clone $this;
+
+        if ('' === $needle) {
+            $str->string = '';
+        } elseif (false === $needle = normalizer_normalize($needle)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        } elseif ($this->ignoreCase) {
+            $i = grapheme_strripos($this->string, $needle);
+            $str->string = false === $i ? '' : grapheme_substr($str->string, $i + ($includeNeedle ? 0 : grapheme_strlen($needle)));
+        } else {
+            $i = strrpos($this->string, $needle);
+            $str->string = false === $i ? '' : substr($str->string, $i + ($includeNeedle ? 0 : \strlen($needle)));
+        }
+
+        return $str;
+    }
+
+    public function append(string $suffix): parent
+    {
+        if ('' !== $suffix && false === $suffix = normalizer_normalize($this->string.$suffix)) {
```

Comment:
is normalizing after concatting intended? i tend to believe both strings are independent, whereas currently it can lead to combined characters

Reply:
grapheme strings are always normalizer, so that yes, this is on purpose.
g('e')->append(combining-accent) should be equals to g('é')